### PR TITLE
foblo+reptile

### PR DESCRIPTION
**Pull request by Jiayi9999**
Merge FOBLO (clean) for spidr-adapt
- Logics update of MetaUpdater Class: 
   - FOBLO and Reptile are children classes of `MetaUpdater`.
   - FOBLO and Reptile share the same `meta_train.py`. 
     - dataloader:  switching from SSL to SL after all inner steps
     - loss weight: switching from SSL to SL after all inner steps
     - learning_rate scheduler:  cyclic (reptile) and cyclic_dual_loss(foblo)

   
- Updated configs files (consistent with our best model). 
   -  added `foblo_ssl.toml`, `foblo_ssl+sl.toml`, `reptile_ssl.toml` and `reptile_ssl+sl.toml`. The main difference is:
      - FOBLO: `inner_step = 1800`, `scheduler = "cyclic_dual_loss"`, `beta = 0.01 / 0.1`
      - Reptile:  `inner_step = 2000`, `scheduler = "cyclic"`, `beta = 0.1`
   - added `meta_update.task_interval` as a config to set up the training steps for each task.
- Important updates of loading init_ckpt in `models/utils.py`
   -   when init the model from init_ckpt, we should use the current `cfg` rather than the init_ckpt's config `json_cfg["model"]`or the default config `config_class()`. Otherwirse, the model hyperparameters in the `.toml` won't work! Especially for `supervised_layer=6/8`, `ema_start_decay=0.999/1.0`,  `ema_end_decay=0.9999/1.0`
   -  when load model parameters, `strict=False` avoids the mismatch of init_ckpt[SSL/SL] and Reptile, in terms of SL layers.

Test Plan: 

> python -m spidr_adapt ~/spidr-adapt/configs/foblo_ssl+sl.toml -A devai -N 2 -G 8 -c 8 -q h200_devai_high -t 4320 --dump ~/submitit

> python -m spidr_adapt ~/project/spidr-adapt/configs/reptile_ssl+sl.toml -A devai -N 2 -G 8 -c 8 -q h200_devai_high -t 4320 --dump ~/submitit

### CODE REVIEW

File: configs/spidr_adapt_foblo.toml [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+model_type = "spidr_reset"
+compile = false
+save_interval = 10000
+keep_latest = 2
+workdir = "/checkpoint/metaspeechssl/jiayishen"
+wandb_mode = "offline"
+wandb_project = "spidr_adapt_codebase"
+wandb_name = "initckpt=sl_madapt-foblo"
+init_ckpt = "/datasets/metaspeechssl/shared/models/spidr_checkpoints/spidr_phoneclass_interleave=10_suplayer=8_vp19_vadcut/step_400000.pt"
+# wandb_name = "initckpt=ssl_madapt-foblo"
```
Comment (mahiluthra) [inline configs/spidr_adapt_foblo.toml]: remove these comments!

File: configs/spidr_adapt_reptile.toml [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+model_type = "spidr_reset"
+compile = false
+save_interval = 10000
+keep_latest = 2
+workdir = "/checkpoint/metaspeechssl/jiayishen"
+wandb_mode = "offline"
+wandb_project = "spidr_adapt_codebase"
+wandb_name = "initckpt=sl_madapt-reptile"
+init_ckpt = "/datasets/metaspeechssl/shared/models/spidr_checkpoints/spidr_phoneclass_interleave=10_suplayer=8_vp19_vadcut/step_400000.pt"
+# wandb_name = "initckpt=ssl_madapt-reptile"
```
Comment (mahiluthra) [inline configs/spidr_adapt_reptile.toml]: remove comments

File: src/spidr_adapt/config.py [hunk 3]
```diff
@@ -95,7 +95,7 @@ class OptimizerConfig:
     rsqrt_timescale: int = 10_000
     rsqrt_shift: int = 0
 
-    scheduler: Literal["tristage", "cosine", "rsqrt", "constant", "cyclic"] = "tristage"
+    scheduler: Literal["tristage", "cosine", "rsqrt", "constant", "cyclic", "cyclic2"] = "tristage"
```
Comment (mahiluthra) [inline src/spidr_adapt/config.py]: can we have a better name than "cyclic2"? maybe cyclic_dual_loss?

File: src/spidr_adapt/data/dataset.py [hunk 4]
```diff
@@ -530,14 +530,13 @@ def __init__(self, cfg: Config, epoch: int, *, supervised_epoch: int = 0) -> Non
...
-                cfg.data.get(BatchType.SUPERVISED.name.lower(), None) if isinstance(cfg.data, dict) else None
-            )
-            assert supervised_data_cfg, "Supervised interleaved training requires supervised data in the config."
-            self.supervised_loader, self.supervised_loader_iter = self.initialize_loaders(supervised_data_cfg)
-            self.supervised_epoch = supervised_epoch
-            self.start_new_epoch(BatchType.SUPERVISED)
+        supervised_data_cfg = (
+            cfg.data.get(BatchType.SUPERVISED.name.lower(), None) if isinstance(cfg.data, dict) else None
+        )
+        assert supervised_data_cfg, "Supervised interleaved training requires supervised data in the config."
```
Comment (mahiluthra) [inline src/spidr_adapt/data/dataset.py]: this assert will fail when using reptile since it is not wrapped within an 'if' loop and we dont instantiate a supervised dataset for reptile.

What you can do instead:
-- remove the assert statement from here.
-- before line 537, add an if statement: `if supervised_data_cfg:`
-- similarly, before line 581, add an if statement: `if self.supervised_data_cfg:`

File: configs/spidr_adapt_reptile.toml [hunk 5]
```diff
@@ -0,0 +1,46 @@
+[data]
+ssl.manifest = "/datasets/metaspeechssl/mahiluthra/split_vad_audio_vp20/vox19_train.csv"
+ssl.by_lang = true
+ssl.lang_task_chunk_duration = 36000
+supervised.manifest = "/datasets/metaspeechssl/mahiluthra/supervised_training_newvp19_50h_train_audio.csv"
```
Comment (mahiluthra) [inline configs/spidr_adapt_reptile.toml]: for reptile, supervised dataset is not needed. so remove these lines instantiating supervised dataset

File: src/spidr_adapt/meta_train.py [hunk 6]
```diff
@@ -131,7 +149,9 @@ def meta_train(cfg: Config) -> None:  # noqa: PLR0914, PLR0915, C901
         loader = get_meta_train_dataloader(
             cfg, step, epoch, supervised_epoch, metalearning_size, metalearning_rank, num_tasks
         )
-        loader.set_task(step, cfg.meta_update.inner_step)
+        logger.info("************** New episode dataset is starting right now at step %s", step)
```
Comment (mahiluthra) [inline src/spidr_adapt/meta_train.py]: Lets remove this logger? We already have a logger statement called by the set_task function below.

File: src/spidr_adapt/meta_train.py [hunk 7]
```diff
@@ -149,13 +169,26 @@ def meta_train(cfg: Config) -> None:  # noqa: PLR0914, PLR0915, C901
         profiler = stack.enter_context(profiler_context(cfg.run.dir / "trace.html" if is_main else None))
         pbar = stack.enter_context(tqdm(total=cfg.optimizer.max_steps, initial=step, disable=not is_main))
         while step < cfg.optimizer.max_steps:
-            batch = loader.load_batch_data(BatchType.SSL)
+            loss_weights, batch_type = get_loss_weights_and_batch_type_reset(
+                step, cfg.meta_update.task_interval, cfg.meta_update.inner_step
+            )
+            # removing after debugging
```
Comment (mahiluthra) [inline src/spidr_adapt/meta_train.py]: remove these debugging lines?

File: src/spidr_adapt/meta_train.py [hunk 8]
```diff
@@ -87,14 +87,28 @@ def get_meta_updater(meta_update_cfg: MetaUpdateConfig, model: SpidRWithReset) -
...
-        # TODO: Jiayi to implement FOBLO
-        raise ValueError("FOBLO not implemented yet")
+        meta_updater = FOBLO(model=model, beta=meta_update_cfg.beta, task_interval=meta_update_cfg.task_interval)
     else:
         raise ValueError(f"Unknown meta_update method: {meta_update_cfg.method}")
     return meta_updater
 
 
-def meta_train(cfg: Config) -> None:  # noqa: PLR0914, PLR0915, C901
+def get_loss_weights_and_batch_type_reset(
```
Comment (mahiluthra) [inline src/spidr_adapt/meta_train.py]: rename this to get_meta_loss_weights_and_batch_type

As part of this, take meta_updater as input and add something like this:
```
if isinstance(meta_updater, Reptile):
          return  LossWeightDefinition(ssl=1.0, supervised=0.0), BatchType.SSL
```

This is just to make it explicit that in reptile, we do not perform supervision

File: src/spidr_adapt/meta_train.py [hunk 9]
```diff
@@ -172,18 +205,25 @@ def meta_train(cfg: Config) -> None:  # noqa: PLR0914, PLR0915, C901
             scheduler.step()
             step += 1
             ema_decay = model.update_ema(step)
-            if step % cfg.meta_update.inner_step == 0:
-                # TODO: Jiayi to implement forward of FOBLO
-                meta_updater(model)  # meta_train update after inner loop
-
-                # reset task after inner loop
+            if cfg.meta_update.method == "reptile":
```
Comment (mahiluthra) [inline src/spidr_adapt/meta_train.py]: this is a complicated branch.

Instead simplify it to just the following:
```
if step % cfg.meta_update.task_interval == 0:
          meta_updater(model)
if step % cfg.meta_update.task_interval == cfg.meta_update.inner_step:
          meta_updater.perform_inner_loop(model)
```

Now, in FOBLO class, rename `update_theta_inner` --> `perform_inner_update`
And in the parent MetaUpdate class, add a function called `perform_inner_update` with `pass`. Reptile will inherit from this function and do nothing during the inner loop.

File: src/spidr_adapt/metalearning/foblo.py [hunk 10]
```diff
@@ -1,26 +1,63 @@
...
         This class is generic and works with any PyTorch Module.
         """
-        # TODO: Jiayi to add FOBLO init
         super().__init__(model=model, beta=beta)
+        self.task_interval = task_interval
+        self.central = copy.deepcopy(model).eval()
+        self.theta_inner = copy.deepcopy(model).eval()
+
+    @torch.no_grad()
+    def update_theta_inner(self, model: nn.Module) -> None:
```
Comment (mahiluthra) [inline src/spidr_adapt/metalearning/foblo.py]: rename to `perform_inner_update`

File: src/spidr_adapt/metalearning/foblo.py [hunk 11]
```diff
@@ -1,26 +1,63 @@
...
-        """Distributed FOBLO.
+    def __init__(self, *, model: nn.Module, beta: float, task_interval: int) -> None:
+        """Distributed Reptile.
 
         This class is generic and works with any PyTorch Module.
         """
-        # TODO: Jiayi to add FOBLO init
         super().__init__(model=model, beta=beta)
+        self.task_interval = task_interval
+        self.central = copy.deepcopy(model).eval()
```
Comment (mahiluthra) [inline src/spidr_adapt/metalearning/foblo.py]: this line 26 can be removed since its inherited from the parent

File: src/spidr_adapt/models/spidr_reset.py [hunk 12]
```diff
@@ -55,13 +55,15 @@ class SpidRWithReset(SpidR):
 
     def __init__(self, cfg: SpidRWithResetConfig) -> None:
         super().__init__(cfg)
-        self.reset_interval = cfg.reset_interval
         self.adapt_heads = cfg.adapt_heads
 
+    def get_task_interval(self, task_interval: int) -> None:
```
Comment (mahiluthra) [inline src/spidr_adapt/models/spidr_reset.py]: rename this to set_task_internal

File: src/spidr_adapt/models/spidr_reset.py [hunk 13]
```diff
@@ -72,18 +74,27 @@ def get_ssl_loss(
...
         target_ppl, pred_ppl = torch.zeros((), device=device), torch.zeros((), device=device)
+
         if should_reset:
-            logger.info("Reset codebook and heads at step %s", self.current_step.item())
+            logger.info("Active Forgetting on codebook and heads at step %s", self.current_step.item())
             reset_codebooks_and_heads(self)
-        for i, (y, target) in enumerate(zip(interm, targets, strict=True)):
-            onehot_target = self.codebooks[i](target)
-            if should_reset and self.adapt_heads:
+            for i, (y, target) in enumerate(zip(interm, targets, strict=True)):
```
Comment (mahiluthra) [inline src/spidr_adapt/models/spidr_reset.py]: why this change? the previous if loop structure was simpler with less repetition

File: src/spidr_adapt/models/utils.py [hunk 14]
```diff
@@ -13,34 +14,40 @@
...
+            warnings.warn("Config file not found when loading checkpoint. Using default config.", stacklevel=2)
+            cfg = config_class()
+
     instance = model_class(cfg)
     state_dict = torch.load(path, map_location="cpu", weights_only=True)
     if "model" in state_dict:
         state_dict = state_dict["model"]
     consume_prefix_in_state_dict_if_present(state_dict, "module.")
-    instance.load_state_dict(state_dict)
+    instance.load_state_dict(state_dict, strict=False)  # avoid mismatch of init_ckpt[SSL/SL] and Reptile, SL_layers.
```
Comment (mahiluthra) [inline src/spidr_adapt/models/utils.py]: Add in a print of mismatched dictionary keys (like we have in metasr_internal)

File: src/spidr_adapt/optimizer.py [hunk 15]
```diff
@@ -53,19 +51,54 @@ def upper_envelope(self, cycle_index: int) -> list[float]:
...
-                + (max_lr - min_lr) * (self.last_epoch % self.reset_interval) / self.cfg.within_cycle_warmup_steps
+                + (max_lr - min_lr) * (self.last_epoch % self.task_interval) / self.cfg.within_cycle_warmup_steps
                 for _ in self.opt.param_groups
             ]
-        return [max_lr for _ in self.opt.param_groups]  # hold phase
+        return [max_lr for _ in self.opt.param_groups]
 
 
-def build_scheduler(opt: Optimizer, cfg: OptimizerConfig, reset_interval: int | None) -> LRScheduler:
+class CyclicLR2(CyclicLR):
```
Comment (mahiluthra) [inline src/spidr_adapt/optimizer.py]: Rename class

File: configs/spidr_adapt_reptile.toml (hunk 2 above)
Comment (Jiayi9999) [reply configs/spidr_adapt_reptile.toml]: they are not comments, but different setups for runs with different inits. It could be better to keep them?

File: configs/spidr_adapt_foblo.toml (hunk 1 above)
Comment (Jiayi9999) [reply configs/spidr_adapt_foblo.toml]: they are not comments, but different setups for runs with different inits. It could be better to keep them?

Comment (mahiluthra) [reply configs/spidr_adapt_foblo.toml]: When committing this code, the init_ckpt will have to be changed to `"/path/to/initialization/checkpoint.pt"` anyway. That is, we will not be providing specific checkpoint names.

File: configs/spidr_adapt_reptile.toml (hunk 2 above)
Comment (mahiluthra) [reply configs/spidr_adapt_reptile.toml]: Same as above!

File: configs/spidr_adapt_foblo.toml (hunk 1 above)
Comment (mahiluthra) [reply configs/spidr_adapt_foblo.toml]: An alternative is to include six configs: multi-task-ssl, multi-task-sl, reptile-ssl, reptile-sl, foblo-ssl, foblo-sl.

Each config can have: init_ckpt either be `null`, `"/path/to/ssl/initialization/checkpoint.pt"` or `"/path/to/ssl_sl/initialization/checkpoint.pt"`

What do you think?

File: configs/spidr_reptile_test.toml [hunk 16]
```diff
@@ -1,18 +1,18 @@
 [run]
-workdir = "/checkpoint/metaspeechssl/mahiluthra"
```
Comment (mahiluthra) [inline configs/spidr_reptile_test.toml:2]: can you remove these changes? I will be deleting these test.toml files before pushing the code!

File: src/spidr_adapt/meta_train.py [hunk 17]
```diff
@@ -140,7 +153,10 @@ def meta_train(cfg: Config) -> None:  # noqa: PLR0914, PLR0915, C901
         ddp_model = DistributedDataParallel(
             model, device_ids=[device.index], find_unused_parameters=True, process_group=pg_ddp()
         )
-        meta_updater = get_meta_updater(cfg.meta_update, model)
+
+        meta_updater = META_UPDATER_CLASSES[cfg.meta_update.method](
```
Comment (mahiluthra) [inline src/spidr_adapt/meta_train.py:157]: nice!

File: src/spidr_adapt/models/spidr_reset.py (hunk 13 above)
Comment (mahiluthra) [reply src/spidr_adapt/models/spidr_reset.py]: @Jiayi9999 can we leave the previous loop structure?

It seems like with the current structure lines 93-98 are repeated. Let me know if i am missing something!

**Review by mahiluthra — COMMENTED** [review]
Thanks Jiayi! This is looking much cleaner 👯 

I have only two quick requests: (1) revert loop structure of spidr_reset (2) leave in the test.toml files for now. Thanks!
